src/topology/homotopy/basic.lean

```diff
@@ -0,0 +1,299 @@
+/-
+Copyright (c) 2021 Shing Tak Lam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Shing Tak Lam
+-/
+
+import topology.unit_interval
+import topology.algebra.ordered.proj_Icc
+import topology.continuous_function.basic
+import topology.compact_open
+
+/-!
+# Homotopy between functions
+
+In this file, we define a homotopy between two functions `f₀` and `f₁`. For this, we use a more
+general `homotopy_with`, which is inspired by the formalisation of homotopy between functions in
+HOL-Analysis. With this, we get `homotopy` as a special case of `homotopy_with`.
+
+## Definitions
+
+* `homotopy_with f₀ f₁ P` is the type of homotoies between `f₀` and `f₁`, where the intermediate
+  maps satisfy the predicate `P : C(X, Y) → Prop`.
+* `homotopy_with.refl` is the homotopy between `f₀` and `f₀`.
+* `homotopy_with.symm F` is a defined by reversing the homotopy `F`
+* `homotopy_with.trans F G`, where `F : homotopy f₀ f₁`, `G : homotopy f₁ f₂` is a `homotopy f₀ f₂`
+  defined by putting the first homotopy on `[0, 1/2]` and the second on `[1/2, 1]`.
+* `homotopy f₀ f₁` is defined to be `homotopy_with f₀ f₁ (λ f, true)`.
+* `homotopy.refl` is a variant of `homotopy_with.refl` with the proof of `P f` filled in.
+* `homotopy_rel f₀ f₁ S` is defined to be
+  `homotopy_with f₀ f₁ (λ f, ∀ x ∈ S, f x = f₀ x ∧ f x = f₁ x)`. That is, a homotopy between `f₀`
+  and `f₁` which is fixed on the points in `S`.
+* `homotopy_rel.refl` is a variant of `homotopy_with.refl` with the proof of `P f` filled in.
+
+## References
+
+- [HOL-Analysis formalisation](https://isabelle.in.tum.de/library/HOL/HOL-Analysis/Homotopy.html)
+-/
+
+noncomputable theory
+
+universes u v
+
+variables {X : Type u} {Y : Type v} [topological_space X] [topological_space Y]
+
+open_locale unit_interval
+
+namespace continuous_map
+
+/--
+The type of homotopies between two functions, where the intermediate maps satisfy the predicate
+`P : C(X, Y) → Prop`.
+-/
+structure homotopy_with (f₀ f₁ : C(X, Y)) (P : C(X, Y) → Prop) extends C(I × X, Y) :=
+(to_fun_zero : ∀ x, to_fun (0, x) = f₀ x)
+(to_fun_one : ∀ x, to_fun (1, x) = f₁ x)
+(prop' : ∀ t, P ⟨λ x, to_fun (t, x),
+  continuous.comp continuous_to_fun (continuous_const.prod_mk continuous_id')⟩)
+
+namespace homotopy_with
+
+section
+
+variables {f₀ f₁ : C(X, Y)} {P : C(X, Y) → Prop}
+
+instance : has_coe_to_fun (homotopy_with f₀ f₁ P) := ⟨_, λ F, F.to_fun⟩
+
+lemma coe_fn_injective : @function.injective (homotopy_with f₀ f₁ P) (I × X → Y) coe_fn :=
+begin
+  rintros ⟨⟨F, _⟩, _⟩ ⟨⟨G, _⟩, _⟩ h,
+  congr' 2,
+end
+
```

Comment:
If you add
```

/-- See Note [custom simps projection]. We need to specify this projection explicitly in this case,
because it is a composition of multiple projections. -/
def simps.apply (h : homotopy_with f₀ f₁ P) : I × X → Y := h

initialize_simps_projections homotopy_with (to_continuous_map_to_fun → apply, -to_continuous_map)
```
here, hopefully you can use `@[simps]` below to autogenerate various lemmas.

Reply:
Thanks!